core/domain/exp_jobs_one_off.py

```diff
@@ -216,69 +260,148 @@ def map(item):
-
-                            # Replacing trig functions that have format which is
-                            # incompatible with the validations.
-                            for invalid_trig_fn, valid_trig_fn in (
-                                    inverse_trig_fns_mapping.items()):
-                                rule_input = rule_input.replace(
-                                    invalid_trig_fn, valid_trig_fn)
-
-                            validity = 'Invalid'
-                            if is_valid_math_expression(rule_input):
-                                validity = 'Valid Expression'
+                            rule_input = _clean_math_expression(rule_input)
+
+                            validity = _INVALID_EXPR
+                            if is_valid_math_expression(
+                                    rule_input, algebraic=True):
+                                validity = _VALID_ALG_EXPR
+                            elif is_valid_math_expression(
+                                    rule_input, algebraic=False):
+                                validity = _VALID_NUM_EXPR
+                            elif is_valid_math_equation(rule_input):
+                                validity = _VALID_MATH_EQN
+
+                            if validity != _INVALID_EXPR:
+                                types_of_inputs.add(validity)
+
+                            if len(types_of_inputs) > 1:
+                                yield ('ERROR', (
+                                    'The exploration with ID: %s and state '
+                                    'name: %s contains inputs that correspond '
+                                    'to multiple different types: %s. Please '
+                                    'resolve this before running the upgrade '
+                                    'job.' % (
+                                        item.id, state_name,
+                                        ', '.join(list(types_of_inputs)))))
+
+                            output_values = '%s %s: %s' % (
+                                item.id, state_name, rule_input)
+                            yield (validity, output_values.encode('utf-8'))
+
+    @staticmethod
+    def reduce(key, values):
+        if key.startswith('Valid'):
+            yield (key, values[:VALID_MATH_INPUTS_YIELD_LIMIT])
+        else:
+            if key == _INVALID_EXPR and len(values) != 0:
+                yield ('ERROR', (
+                    'There are some invalid inputs that need to be resolved '
+                    'before running the upgrade job. The invalid inputs '
+                    'are as follows:'))
+            yield (key, values)
+
+
+class MathExpressionUpgradeOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that produces a list of explorations that use the MathExpressionInput
+    along with the validity and type (expression/equation) of the inputs present
+    in the exploration.
```

Comment:
It also upgrades the exploration as well, right?

Reply:
Ah yes, I've updated the comment.